### DISCUSSION TITLE
Error in custom sql `order by` only when `where` clause present

### DISCUSSION THREAD

**Discussion by brsanthu** [eclipse-langium/langium#1480] [Q&A] [answered]
We have a custom sql dsl that supports subset of sql as outlined in [playground](https://langium.org/playground?grammar=OYJwhgthYgBAwgZwI4BsBQ6CmA7ALiAJ4IqoCyA9gCZaqwCWOAZliIiWpTagFzqwDYiPGDxYIuPIgDUAXgDKIsRPwAqdAG5Mi0eMkNmrdjuWS%2BgjnQDkGqwH5NmJGnOCriWlgDGeK-0FM9LRUMgqePgBiQahUAgAUVgA0VrCBwaHy4XhRwQIAlOoWVkwgFBApJWUA8gBGAFbeeLIAcpBYsHGVEACCqPRgiLIAkgAieQ4WAO4AFqxYsgDqsyBY8KhgAK4eE4IJFCA0IClWNYQp%2B4cAQoRyVQes10PK8Unn9yDXt%2B%2BPygXjjug7ldCE9xK4BIgKBsQF55nEcjFYAAfWAAJVoCKoeQ6VHoKx89AoOFkCQGXhSKKsNEQ5NgeX%2BWnQSzma02HgMLDYsGZK1ZWyw4NgVhmcxScSwAA8AA5sOQ8rAAUWlcA6kploXlSpldNUdIEKISOAovlgatli2WqyJuLwhJw%2BQBmuVHKM3MtWpAgruHoB3udjE57CdMsF3RwVA9HQA3i64H6ZQA6VBYJhNLzQlb4AC%2BQv2KRA9GA0yaYYjzp1ANLkYDruDnv8AgACgXoERI3EYzW4FXlUmU2mM5Ic1YwOH84Xi7Jm-RW4RIxXGdPZ9XDFy64KpTBcHZZFYEm65pHN5md1Y8ilkQfedb6LaiY7LfAb3f7V2g4-n3bBZDobCSZjL3RVBMWxOIjTwU9wJScZYAoKVWFEfYLTmKp4PAPB9gEAA3MBUA2eZ5QANVw-CHzmYi8PaN8rywCj8MFRQCxwYA6KwAFGMYFiSKo1djAITjWMFHDKIUIZmgAcQAGQVAB9eQABVUTE8T2KyADqMyZNImiKhBQAlEFnoGJ9O5IyqCAzEATwQh4Jo1CEIwuBZCFXdLysABCVzKQAHgpIVvK8oUAD4-KsILAqsPoAGssFCxg4pwLw8OpULJSSjYUq0eV7PQzCVjwaEcHfFC0MQ%2BsigizzQt8tyAtCkK3PC0Lotitz4raxLkqwRBUoldKUoBdTeNgTFBRwNoWjaAELJ0saJqA1oJEdMyhsDUzjNmhtYHGiRFiGSSRhk0ZlpiGbcmowyNuCObdoARQ2XD6DSKhLqoRa2MZIYqDuV6vtgfLCvYYQmOAQUFn2w7RkvY7GXux7nte97-qwAqQCKoR%2BOYwUvp%2Bsy-oSBMUhxkBftiPJpGm2gkYBtGgcx0GtqRgmUiR8mAWplHAYxkHBSsC5WD8mHMDEEAIEYXDuQho6RmR1H0eBzieFgAB6OI4gAbRksAAFoAC9um1gAtAAGbWAE5VAAXXJgAdVQNa1vWDZN82rfJvIkTiO2HZ1-WjdNi3rekD21c133nYDt3pDtj27eVrQRbF8a6FGJXlbDp2jctjO-Zds3LdUeP0ET8W6HkZTpLkxTlNlrmFaxgAiKwG46KwbZtlIE0vWB3ISdu-Kblv6VUQeBEZaZ6CoGh7RL5PYDISSZPgKoyDIBVmnktObeVu31ZtxAbfkAu7Dt7ei4nqfcFgWeJfkRfl9X9fN5V7ft-VgA9G2cBtkAC6L9AL7T2vqwJOEsFjyFrrTbmit9yhRtiaSkP84E4GgtILQQA&content=M4UwNiDGAuAECWATWAzATgewLawIaUgwFcA7aAKAzURDVgCMBPWE3LEAbnPNAhgWTpseAsTLkA7gAtaIAbAC8sAOSt2yytVoNmazuSA). 

As you can see, the `order by name` errors when there is a `where` clause otherwise it parses fine.

One peculiar thing with this dsl is we could have table name conflicts with keywords. For ex., `order` so we have specified `KeywordOrId` Rule which allows specified keyword variables like this. Error also goes away if we remove `'order' |` from the `Name` rule.

```
Name returns string:
    'order' | ID
;
```

Any idea what I'm doing wrong?

**Comment by msujew** [comment, marked as answer]
Hey @brsanthu,

you're basically hitting an issue in our [parser implementation](https://github.com/TypeFox/chevrotain-allstar/issues/1). Mainly, a `WhereExpr` can be followed by any other amount of `WhereExpr`. Since `order` is handled as a `Name`, it will think that there's a new `WhereExpr` starting. While there's a fix that includes rewriting the grammar to lift different rules to the same level, note that this isn't valid SQL syntax, as multiple `WhereExpr` should be separated by `AND` or `OR`. You should adjust your grammar. See also [langium-sql](https://github.com/TypeFox/langium-sql/blob/c20515756c30779a34c73d81238da8ec94cc1ab3/packages/langium-sql/src/sql.langium#L158-L159) for a mostly complete implementation of DQL. That should fix the issue by itself.

Also note that SQL forbids keywords as identifiers, you should use string escaping instead (also implemented in langium-sql).

**Reply by brsanthu** [reply]
Thank you, now I see the issue with how `WhereExpr` is structured. Will fix it and retry.

Did you identify the root cause by looking at the grammar or by going through some intermediate generated json or js file? Just trying to udnerstand the flow of troubleshooting for such errors.

**Reply by msujew** [reply]
> Did you identify the root cause by looking at the grammar or by going through some intermediate generated json or js file? Just trying to udnerstand the flow of troubleshooting for such errors.

It's mostly based on experience by looking at the grammar. I've seen that issue at least a dozen times in the past few years. The generated code from Langium is mostly useless to understand parser issues, as the parser is contructed from the grammar directly without an intermediate representation.
